Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 20807004c45cfa8154bccb05bea8c83ec5bc1748
Number of Error(s): 5
/mnt/disk1/home/vamosax/smoketest/PR-11142/HPCC-Platform/esp/services/ws_workunits/ws_workunitsService.cpp:4030:14: error: \xe2\x80\x98struct IEspWUDetailsMetaResponse\xe2\x80\x99 has no member named \xe2\x80\x98setStatistics\xe2\x80\x99

[Build error:](https://gist.github.com/8de97e24d603d4df9b6c21ca22308cd1)
Build log [Build log](https://gist.github.com/HPCCSmoketest/8de97e24d603d4df9b6c21ca22308cd1#file-result-18-05-14-09-22-50-log)

--- a/common/workunit/wuattr.cpp
+++ b/common/workunit/wuattr.cpp
@@ -312,6 +312,12 @@ bool isListAttribute(WuAttr kind)
     return (info.multiKind != WaNone) && (info.multiKind == kind);
 }
 
+bool isMultiAttribute(WuAttr kind)
+{
+    const WuAttrInfo & info = attrInfo[kind-WaNone];
+    return (info.singleKind != WaNone) && (info.singleKind == kind);
+}
+
 WuAttr getListAttribute(WuAttr kind)
 {
     const WuAttrInfo & info = attrInfo[kind-WaNone];

--- a/common/workunit/wuattr.hpp
+++ b/common/workunit/wuattr.hpp
@@ -134,7 +134,7 @@ enum WuAttr : unsigned
     WaCriticalSection,
     WaMax
 };
-
+inline WuAttr & operator++(WuAttr & x) { assert(x<WaMax); x = (WuAttr)(x+1); return x; }
 extern WORKUNIT_API const char * queryWuAttributeName(WuAttr kind);
 extern WORKUNIT_API WuAttr queryWuAttribute(const char * kind, WuAttr dft);
 extern WORKUNIT_API const char * queryAttributeValue(IPropertyTree & src, WuAttr kind, StringBuffer & scratchpad); // may return pointer to scratchpad if necessary
@@ -146,6 +146,7 @@ extern WORKUNIT_API void setAttributeValueBool(IPropertyTree & tgt, WuAttr kind,
 extern WORKUNIT_API void setAttributeValueInt(IPropertyTree & tgt, WuAttr kind, __int64 value);
 
 extern WORKUNIT_API bool isListAttribute(WuAttr kind);
+extern WORKUNIT_API bool isMultiAttribute(WuAttr kind);
 extern WORKUNIT_API WuAttr getListAttribute(WuAttr kind);
 
 #endif

--- a/esp/scm/ws_workunits.ecm
+++ b/esp/scm/ws_workunits.ecm
@@ -2148,15 +2148,37 @@ ESPResponse [exceptions_inline] WUDetailsResponse
 };
 
 // ----------------------------------------------------------------------------------
+ESPenum WUDetailsAttrValueType : string
+{
+    Single("Single"),
+    List("List"),
+    Multi("Multi"),
+};
+
+ESPStruct WUDetailsMetaProperty
+{
+    string Name;
+    ESPenum WUDetailsAttrValueType ValueType;
+};
+
+ESPStruct WUDetailsActivityInfo
+{
+    unsigned Kind;
+    string Name;
+    boolean IsSink;
+    boolean IsSource;
+};
+
 ESPRequest WUDetailsMetaRequest
 {
 };
 
 ESPResponse [exceptions_inline] WUDetailsMetaResponse
 {
-    ESParray<string, Property> Properties;
+    ESParray<ESPStruct WUDetailsMetaProperty, Property> Properties;
     ESParray<string, ScopeType> ScopeTypes;
     ESParray<string, Measure> Measures;
+    ESParray<ESPStruct WUDetailsActivityInfo, Activity> Activities;
 };
 
 ESPStruct [nil_remove] WUEclDefinitionActionResult
@@ -2286,7 +2308,7 @@ ESPservice [
     ESPmethod [cache_seconds(60), min_ver("1.57")] WUGetArchiveFile(WUGetArchiveFileRequest, WUGetArchiveFileResponse);
     ESPmethod [cache_seconds(60), min_ver("1.61")] WUGetNumFileToCopy(WUGetNumFileToCopyRequest, WUGetNumFileToCopyResponse);
     ESPmethod [min_ver("1.71")] WUDetails(WUDetailsRequest, WUDetailsResponse);
-    ESPmethod [min_ver("1.71")] WUDetailsMeta(WUDetailsMetaRequest, WUDetailsMetaResponse);
+    ESPmethod [cache_seconds(600),min_ver("1.71")] WUDetailsMeta(WUDetailsMetaRequest, WUDetailsMetaResponse);
     ESPmethod [min_ver("1.72")] WUEclDefinitionAction(WUEclDefinitionActionRequest, WUEclDefinitionActionResponse);
 };
 

--- a/esp/services/ws_workunits/ws_workunitsService.cpp
+++ b/esp/services/ws_workunits/ws_workunitsService.cpp
@@ -42,6 +42,7 @@
 #include "roxiecontrol.hpp"
 
 #include "deftype.hpp"
+#include "thorcommon.hpp"
 #include "thorxmlwrite.hpp"
 #include "fvdatasource.hpp"
 #include "fvresultset.ipp"
@@ -4014,39 +4015,66 @@ bool CWsWorkunitsEx::onWUDetailsMeta(IEspContext &context, IEspWUDetailsMetaRequ
 {
     try
     {
-        StringArray properties;
-        for (unsigned i=StatisticKind::StKindAll+1; i<StatisticKind::StMax;++i)
+        IArrayOf<IEspWUDetailsMetaProperty> properties;
+        for (unsigned sk=StKindAll+1; sk<StMax;++sk)
         {
-            const char * s = queryStatisticName((StatisticKind)i);
+            const char * s = queryStatisticName((StatisticKind)sk);
             if (s && *s)
-                properties.append(s);
+            {
+                Owned<IEspWUDetailsMetaProperty> property = createWUDetailsMetaProperty("","");
+                property->setName(s);
+                property->setValueType(CWUDetailsAttrValueType_Single);
+                properties.append(*property.getClear());
+            }
         }
 
-        for (unsigned i=WuAttr::WaAll+1; i<WuAttr::WaMax; ++i)
+        for (WuAttr attr=WaKind; attr<WaMax; ++attr)
         {
-            const char * s = queryWuAttributeName((WuAttr)i);
-            if (s && *s)
-                properties.append(s);
+            Owned<IEspWUDetailsMetaProperty> property = createWUDetailsMetaProperty("","");
+            const char * s = queryWuAttributeName(attr);
+            assertex(s && *s);
+            property->setName(s);
+            if (isListAttribute(attr))
+                property->setValueType(CWUDetailsAttrValueType_List);
+            else if (isMultiAttribute(attr))
+                property->setValueType(CWUDetailsAttrValueType_Multi);
+            else
+                property->setValueType(CWUDetailsAttrValueType_Single);
+            properties.append(*property.getClear());
         }
         resp.setProperties(properties);
 
         StringArray scopeTypes;
-        for (unsigned i=StatisticScopeType::SSTall+1; i<StatisticScopeType::SSTmax; ++i)
+        for (unsigned sst=SSTall+1; sst<SSTmax; ++sst)
         {
-            const char * s = queryScopeTypeName((StatisticScopeType)i);
+            const char * s = queryScopeTypeName((StatisticScopeType)sst);
             if (s && *s)
                 scopeTypes.append(s);
         }
         resp.setScopeTypes(scopeTypes);
 
         StringArray measures;
-        for (unsigned i=StatisticMeasure::SMeasureAll+1; i<StatisticMeasure::SMeasureMax; ++i)
+        for (unsigned measure=SMeasureAll+1; measure<SMeasureMax; ++measure)
         {
-            const char *s = queryMeasureName((StatisticMeasure)i);
+            const char *s = queryMeasureName((StatisticMeasure)measure);
             if (s && *s)
                 measures.append(s);
         }
         resp.setMeasures(measures);
+        IArrayOf<IConstWUDetailsActivityInfo> activities;
+        for (unsigned kind=((unsigned)ThorActivityKind::TAKnone)+1; kind< TAKlast; ++kind)
+        {
+            Owned<IEspWUDetailsActivityInfo> activity = createWUDetailsActivityInfo("","");
+            const char * name = getActivityText(static_cast<ThorActivityKind>(kind));
+            assertex(name && *name);
+            activity->setKind(kind);
+            activity->setName(name);
+            activity->setIsSink(isActivitySink(static_cast<ThorActivityKind>(kind)));
+            activity->setIsSource(isActivitySource(static_cast<ThorActivityKind>(kind)));
+            activities.append(*activity.getClear());
+        }
+        resp.setActivities(activities);
+
     }
     catch(IException* e)
     {

